File: lib/contract/canonical_claims.ex
# lib/contract/canonical_claims.ex — canonical claims (the wire shape) → engine claims (gr-3jd).
#
# The GENERIC half of the ingest split: source-agnostic, contract-driven translation from
# canonical claim maps (docs/CLAIMS_CONTRACT.md — the same shape `POST /v1/claims` accepts) into
# the engine's `%Events.ClaimAsserted{}` structs. Every path into the engine meets here: the
# medipim reference adapter (lib/ingest/claim_mapping.ex) derives these maps from legacy
# envelopes; the Product API's live path (`Api.Writes.claims/1`, gr-ajc) takes them off the wire.
#
# Validation at the seam — DOCUMENTED CHOICE (gr-3jd): the executable contract
# (`ClaimsValidator`) runs in `to_engine/2`, the live-wire entrypoint, and is SKIPPED by
# `to_engine!/2`, the trusted/backfill entrypoint. A backfill batch deliberately exceeds the
# live wire on two axes the validator (correctly, per the spec) rejects:
#
#   * `member_of` claims — produced by the medipim adapter but not yet on the wire contract
#     (open question 1 in docs/CLAIMS_CONTRACT.md; shape below follows its candidate).
#   * temporal fields as unix-second integers (contract C's clock, docs/HISTORY_ENVELOPE.md)
#     instead of ISO dates — the backfill carries the historical `recorded_at` the live
#     contract reserves for the server.

defmodule CanonicalClaims do
  @moduledoc """
  Generic canonical-claims → engine-claims translation (`docs/CLAIMS_CONTRACT.md`).

  Input is a decoded claims batch: a list of string-keyed claim maps in the wire shape. Codes
  are `"scheme:value"` strings (`parse_code/1` / `code_string/1` are the two directions);
  unknown schemes pass through as strings — conservative, never an atom leak.

  Two entrypoints, differing only in whether the executable contract runs at the seam:

    * `to_engine/2` — VALIDATES the batch with `ClaimsValidator` first; `{:ok, claims}` or
      `{:error, errors}` (whole batch, nothing partial). The live-wire path.
    * `to_engine!/2` — no validation, raises on a malformed claim. The trusted path for
      backfill batches, which deliberately exceed the live wire (`member_of` claims and
      unix-second temporal fields — see the header comment).

  Temporal fields: each claim carries its own `"recorded_at"` (backfill), unless the caller
  passes `recorded_at:` (the live path's server-side clock, which clients cannot supply).
  `"valid_from"` is optional — an ISO 8601 date string (live wire) or a unix-second integer
  (backfill flavor) — and defaults to the claim's `recorded_at`.
  """

  @doc """
  Validate the batch against the contract, then translate. `{:ok, [%Events.ClaimAsserted{}]}`
  in batch order, or `{:error, errors}` with `ClaimsValidator`'s per-index findings.
  """
  def to_engine(claims, opts \\ []) do
    case ClaimsValidator.validate(claims) do
      {:ok, _warnings} -> {:ok, to_engine!(claims, opts)}
      {:error, errors} -> {:error, errors}
    end
  end

  @doc """
  Translate WITHOUT validating — the trusted/backfill entrypoint (see the module doc for why a
  backfill batch cannot pass the live-wire validator). Raises on a malformed claim.
  """
  def to_engine!(claims, opts \\ []) when is_list(claims) do
    recorded_at = Keyword.get(opts, :recorded_at)
    Enum.map(claims, &build(&1, recorded_at))
  end

  # ── one claim map → one engine claim ────────────────────────────────────────
  defp build(%{"kind" => "identity", "source" => s, "ref" => ref, "codes" => codes} = m, at) do
    data = %{ref: ref, codes: Enum.map(codes, &code!/1)}
    Substrate.claim(s, :identity, data, valid_from(m, at), recorded_at(m, at))
  end

  defp build(
         %{"kind" => "attribute", "source" => s, "code" => c, "field" => f, "value" => v} = m,
         at
       ) do
    data = %{code: code!(c), field: f, value: v}
    Substrate.claim(s, :attribute, data, valid_from(m, at), recorded_at(m, at))
  end

  defp build(
         %{"kind" => "media", "source" => s, "asset" => a, "target" => t, "uri" => uri} = m,
         at
       ) do
    role = if m["role"] == "primary", do: :primary, else: :secondary
    data = %{asset: {:dam, a}, target: code!(t), role: role, uri: uri}
    Substrate.claim(s, :media, data, valid_from(m, at), recorded_at(m, at))
  end

  defp build(%{"kind" => "grouping", "source" => s, "code" => c, "product" => p} = m, at) do
    Substrate.claim(s, :grouping, %{code: code!(c), product: p}, valid_from(m, at), recorded_at(m, at))
  end

  # member_of is NOT on the live wire yet (open question 1 in docs/CLAIMS_CONTRACT.md); this is
  # its candidate shape: `code` is a member of `collection`/`member` (e.g. brands/211). The
  # collection name is NOT scheme-folded — it is a collection namespace, not a code scheme.
  defp build(
         %{"kind" => "member_of", "source" => s, "code" => c, "collection" => coll, "member" => member} = m,
         at
       ) do
    data = %{member_code: code!(c), collection: {coll, member}}
    Substrate.claim(s, :member_of, data, valid_from(m, at), recorded_at(m, at))
  end

  # ── codes: the two directions of "scheme:value" ─────────────────────────────
  @doc ~s(Parse one "scheme:value" string into an engine {scheme, value} code tuple.)
  def parse_code(raw) when is_binary(raw) do
    case String.split(raw, ":", parts: 2) do
      [scheme, value] when scheme != "" and value != "" ->
        {:ok, {CodeRegistry.engine_scheme(scheme), value}}

      _ ->
        {:error, ~s(code must be "scheme:value", got #{inspect(raw)})}
    end
  end

  def parse_code(raw),
    do: {:error, ~s(code must be a "scheme:value" string, got #{inspect(raw)})}

  @doc ~s"""
  Format an engine `{scheme, value}` code tuple as a `"scheme:value"` wire string — the inverse
  of `parse_code/1` (engine-native scheme atoms round-trip via `CodeRegistry.engine_scheme/1`;
  unknown string schemes pass through, provided they contain no colon).
  """
  def code_string({scheme, value}), do: "#{scheme}:#{value}"

  # ── temporal fields ─────────────────────────────────────────────────────────
  # Caller-supplied recorded_at (live path: the server's clock) wins; otherwise the claim must
  # carry its own (backfill path: contract C's unix seconds).
  defp recorded_at(m, nil), do: Map.fetch!(m, "recorded_at")
  defp recorded_at(_m, at), do: at

  defp valid_from(%{"valid_from" => raw}, _at) when is_binary(raw), do: Date.from_iso8601!(raw)
  defp valid_from(%{"valid_from" => vf}, _at) when not is_nil(vf), do: vf
  defp valid_from(m, at), do: recorded_at(m, at)

  defp code!(raw) do
    {:ok, code} = parse_code(raw)
    code
  end
end

File: php/src/CanonicalClaims.php
<?php

declare(strict_types=1);

namespace Ingot;

/**
 * Canonical claims (the wire shape) → engine claims — ported from `CanonicalClaims`
 * (lib/contract/canonical_claims.ex). The generic, source-agnostic half of the ingest split.
 *
 * Only the trusted/backfill entrypoint `toEngineBang` is ported (the 422156 fold path uses it via
 * ClaimMapping); the live-wire `to_engine/2` validator is out of scope. Codes are "scheme:value"
 * strings — `parseCode`/`codeString` are the two directions.
 */
final class CanonicalClaims
{
    /**
     * Translate WITHOUT validating (the trusted/backfill entrypoint). `$recordedAt` overrides each
     * claim's own recorded_at when supplied (the live path's server clock).
     *
     * @param list<array<string,mixed>> $claims wire-shaped claim maps
     * @return list<array<string,mixed>> engine ClaimAsserted arrays
     */
    public static function toEngineBang(array $claims, mixed $recordedAt = null): array
    {
        $out = [];
        foreach ($claims as $m) {
            $out[] = self::build($m, $recordedAt);
        }

        return $out;
    }

    /**
     * @param array<string,mixed> $m
     * @return array<string,mixed>
     */
    private static function build(array $m, mixed $at): array
    {
        return match ($m['kind']) {
            'identity' => self::buildIdentity($m, $at),
            'edge' => self::buildEdge($m, $at),
            'attribute' => self::buildAttribute($m, $at),
            'media' => self::buildMedia($m, $at),
            'grouping' => self::buildGrouping($m, $at),
            'member_of' => self::buildMemberOf($m, $at),
            default => throw new \InvalidArgumentException('malformed claim kind: '.($m['kind'] ?? '?')),
        };
    }

    /**
     * @param array<string,mixed> $m
     * @return array<string,mixed>
     */
    private static function buildIdentity(array $m, mixed $at): array
    {
        $codes = array_map(self::codeBang(...), $m['codes']);
        $data = ['ref' => $m['ref'], 'codes' => $codes];
        if (isset($m['entity'])) {
            $lane = Lanes::parse($m['entity']);
            if ($lane === null) {
                throw new \InvalidArgumentException('unknown entity lane: '.$m['entity']);
            }
            $data['entity'] = $lane;
        }

        return Substrate::claim($m['source'], 'identity', $data, self::validFrom($m, $at), self::recordedAt($m, $at));
    }

    /**
     * @param array<string,mixed> $m
     * @return array<string,mixed>
     */
    private static function buildEdge(array $m, mixed $at): array
    {
        $relation = Relations::parse($m['relation']);
        if ($relation === null) {
            throw new \InvalidArgumentException('unknown relation: '.$m['relation']);
        }
        $data = ['from' => self::codeBang($m['from']), 'relation' => $relation, 'to' => self::codeBang($m['to'])];

        return Substrate::claim($m['source'], 'edge', $data, self::validFrom($m, $at), self::recordedAt($m, $at));
    }

    /**
     * @param array<string,mixed> $m
     * @return array<string,mixed>
     */
    private static function buildAttribute(array $m, mixed $at): array
    {
        $data = ['code' => self::codeBang($m['code']), 'field' => $m['field'], 'value' => $m['value']];

        return Substrate::claim($m['source'], 'attribute', $data, self::validFrom($m, $at), self::recordedAt($m, $at));
    }

    /**
     * @param array<string,mixed> $m
     * @return array<string,mixed>
     */
    private static function buildMedia(array $m, mixed $at): array
    {
        $role = ($m['role'] ?? null) === 'primary' ? 'primary' : 'secondary';
        $data = ['asset' => ['dam', $m['asset']], 'target' => self::codeBang($m['target']), 'role' => $role, 'uri' => $m['uri']];

        return Substrate::claim($m['source'], 'media', $data, self::validFrom($m, $at), self::recordedAt($m, $at));
    }

    /**
     * @param array<string,mixed> $m
     * @return array<string,mixed>
     */
    private static function buildGrouping(array $m, mixed $at): array
    {
        $data = ['code' => self::codeBang($m['code']), 'product' => $m['product']];

        return Substrate::claim($m['source'], 'grouping', $data, self::validFrom($m, $at), self::recordedAt($m, $at));
    }

    /**
     * @param array<string,mixed> $m
     * @return array<string,mixed>
     */
    private static function buildMemberOf(array $m, mixed $at): array
    {
        $data = ['member_code' => self::codeBang($m['code']), 'collection' => [$m['collection'], $m['member']]];

        return Substrate::claim($m['source'], 'member_of', $data, self::validFrom($m, $at), self::recordedAt($m, $at));
    }

    /**
     * Parse one "scheme:value" string into an engine [scheme, value] code, splitting on the FIRST
     * colon. Returns ['ok', code] or ['error', message].
     *
     * @return array{0: string, 1: mixed}
     */
    public static function parseCode(mixed $raw): array
    {
        if (!is_string($raw)) {
            return ['error', 'code must be a "scheme:value" string'];
        }
        $parts = explode(':', $raw, 2);
        if (count($parts) !== 2 || $parts[0] === '' || $parts[1] === '') {
            return ['error', 'code must be "scheme:value"'];
        }

        return ['ok', [CodeRegistry::engineScheme($parts[0]), $parts[1]]];
    }

    /**
     * Format an engine [scheme, value] code as a "scheme:value" wire string.
     *
     * @param array{0: string, 1: string} $code
     */
    public static function codeString(array $code): string
    {
        return $code[0].':'.$code[1];
    }

    /**
     * @return array{0: string, 1: string}
     */
    private static function codeBang(string $raw): array
    {
        $result = self::parseCode($raw);
        if ($result[0] !== 'ok') {
            throw new \InvalidArgumentException('malformed code: '.$raw);
        }

        return $result[1];
    }

    /** @param array<string,mixed> $m */
    private static function recordedAt(array $m, mixed $at): mixed
    {
        if ($at !== null) {
            return $at;
        }
        if (!array_key_exists('recorded_at', $m)) {
            throw new \InvalidArgumentException('claim missing recorded_at');
        }

        return $m['recorded_at'];
    }

    /** @param array<string,mixed> $m */
    private static function validFrom(array $m, mixed $at): mixed
    {
        if (array_key_exists('valid_from', $m) && is_string($m['valid_from'])) {
            // ISO date string (live wire) — not used on the backfill path, but kept for fidelity.
            return $m['valid_from'];
        }
        if (array_key_exists('valid_from', $m) && $m['valid_from'] !== null) {
            return $m['valid_from'];
        }

        return self::recordedAt($m, $at);
    }
}
